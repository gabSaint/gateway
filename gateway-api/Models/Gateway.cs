using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.ComponentModel.DataAnnotations;
using gateway_api.Contexts;
using gateway_api.Extensions;

namespace gateway_api.Models
{
    public class Gateway
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public string Serial { get; set; }

        [Required]
        public string Name { get; set; }

        [Required]
        public string Address { get; set; }
    }

    public static class GatewaySeed
    {
        public static void InitData(GatewayContext context)
        {
            var rnd = new Random();

            var serials = new[] { "2HJ561", "18K3LO", "09DP4K", "NWE3M4", "NT88M4" };

            var names = new[] { "gateway1", "gateway2", "gateway3", "gateway4", "gateway5", "gateway6" };

            var addresses = new [] {"127.0.0.1", "127.0.0.2", "127.0.0.3", "127.0.0.4", "127.0.0.5" };

            context.Gateways.AddRange(100.Times(x =>
            {
                var serial = serials[rnd.Next(0, 5)];
                var name = names[rnd.Next(0, 6)];
                var address = addresses[rnd.Next(0, 5)];
                var id = $"{x,-3:000}";

                return new Gateway
                {
                    Id = x,
                    Name = name,
                    Serial = serial,
                    Address = address
                };
            }));

            context.SaveChanges();
        }
    }
}
