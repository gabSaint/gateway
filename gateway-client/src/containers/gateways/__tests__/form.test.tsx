import React from "react";
import { ReactWrapper, shallow, ShallowWrapper } from "enzyme";

import Form, {
  FormContent,
  Props as FormGatewaysProps,
  validate,
} from "../form";
import { FormikState } from "formik";
import Gateway from "models/gateway";

describe("FormGateways", () => {
  let mountedGatewayForm: ReactWrapper | ShallowWrapper;
  let props: FormGatewaysProps;

  beforeEach(() => {
    props = {
      gateway: {
        id: 1,
        serial: "24589",
        name: "gateway1",
        address: "127.0.27.1",
      },
      handleSubmit: jest.fn(),
    };
    mountedGatewayForm = shallow(<Form {...props} />);
  });

  it("renders form", () => {
    const formik = mountedGatewayForm.find("Formik");
    expect(formik).toHaveLength(1);
  });
});

describe("FormContent", () => {
  let mountedGatewayFormContent: ReactWrapper | ShallowWrapper;
  let props: FormikState<Gateway>;

  beforeEach(() => {
    props = {
      errors: { serial: "error", name: "error", address: "error" },
      touched: { serial: true, name: true, address: true },
    } as FormikState<Gateway>;
    mountedGatewayFormContent = shallow(<FormContent {...props} />);
  });

  describe("Form", () => {
    it("renders submit button", () => {
      const buttonsubmit = mountedGatewayFormContent.find(
        "button[type='submit']"
      );
      expect(buttonsubmit).toHaveLength(1);
    });

    it("renders cancel button", () => {
      const buttoncancel = mountedGatewayFormContent.find(
        "button[type='button']"
      );
      expect(buttoncancel).toHaveLength(1);
    });

    it("renders serial number input", () => {
      const inputserialnumbers = mountedGatewayFormContent.find(
        "#input-serial"
      );
      expect(inputserialnumbers).toHaveLength(1);
    });

    it("renders name input", () => {
      const inputnames = mountedGatewayFormContent.find("#input-name");
      expect(inputnames).toHaveLength(1);
    });

    it("renders address input", () => {
      const inputaddress = mountedGatewayFormContent.find("#input-address");
      expect(inputaddress).toHaveLength(1);
    });
  });

  describe("Behaviour", () => {
    it("shows error when wrong serial", () => {
      const serial = mountedGatewayFormContent.find(
        "[data-test='show-serial-error']"
      );
      expect(serial).toHaveLength(1);
    });

    it("shows error when wrong name", () => {
      const name = mountedGatewayFormContent.find(
        "[data-test='show-name-error']"
      );
      expect(name).toHaveLength(1);
    });

    it("shows error when wrong address", () => {
      const address = mountedGatewayFormContent.find(
        "[data-test='show-address-error']"
      );
      expect(address).toHaveLength(1);
    });

    describe("Function Validate", () => {
      let correctGateway: any;
      let wrongGateway: any;

      beforeEach(() => {
        correctGateway = validate(
          new Gateway(1, "12345", "gate1", "127.0.0.1")
        );
        wrongGateway = validate(new Gateway(2, "", "", "127.1"));
      });

      it("validates correctly the serial input", () => {
        expect(correctGateway.serial).toBeUndefined();
        expect(wrongGateway.serial).toBeDefined();
      });

      it("validates correctly the name input", () => {
        expect(correctGateway.name).toBeUndefined();
        expect(wrongGateway.name).toBeDefined();
      });

      it("validates correctly the address input", () => {
        expect(correctGateway.address).toBeUndefined();
        expect(wrongGateway.address).toBeDefined();
      });
    });
  });
});
