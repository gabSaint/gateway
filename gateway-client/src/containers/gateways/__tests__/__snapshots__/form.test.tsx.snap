// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGateways renders correctly 1`] = `
<form
  action="#"
  className="ui form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <label
      htmlFor="input-serial"
    >
      Serial Number
    </label>
    <input
      id="input-serial"
      name="serial"
      onBlur={[Function]}
      onChange={[Function]}
      value="24589"
    />
  </div>
  <div
    className="field"
  >
    <label
      htmlFor="input-name"
    >
      Name
    </label>
    <input
      id="input-name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      value="gateway1"
    />
  </div>
  <div
    className="field"
  >
    <label
      htmlFor="input-address"
    >
      IPv4 Address
    </label>
    <input
      id="input-address"
      name="address"
      onBlur={[Function]}
      onChange={[Function]}
      value="127.0.27.1"
    />
  </div>
  <button
    type="submit"
  >
    Submit
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Cancel
  </button>
</form>
`;
